pyemma/coordinates/clustering/include/bits/kmeans_bits.h

```diff
@@ -0,0 +1,328 @@
+    //#pragma omp parallel for
+    for (size_t r = 0; r < np_centers.shape(0); r++) {
+        for (size_t i = 0; i < n_frames; i++) {
+            auto l = parent_t::metric->compute(&data(i, 0), &centers(r, 0));
+            #pragma omp critical
+            {
+                value += l;
+            }
+        }
+    }
+    return value;
+}
+
+template<typename dtype>
+typename KMeans<dtype>::np_array KMeans<dtype>::
+initCentersKMpp(const np_array& np_data, unsigned int random_seed, int n_threads) const {
+    size_t centers_found = 0, first_center_index;
+    bool some_not_done;
+    dtype sum;
+    size_t i, j;
+    size_t dim = parent_t::metric->dim;
+
+    if (np_data.ndim() != 2) {
+        throw std::invalid_argument("input data does not have two dimensions.");
+    }
+
+    if (np_data.shape(1) != dim) {
+        throw std::invalid_argument("input dimension of data does not match the requested metric ones.");
+    }
+
+    size_t n_frames = np_data.shape(0);
+
+    /* number of trials before choosing the data point with the best potential */
+    size_t n_trials = 2 + (size_t) log(k);
+
+    /* allocate space for the index giving away which point has already been used as a cluster center */
+    std::vector<bool> taken_points(n_frames, false);
+    /* candidates allocations */
+    std::vector<int> next_center_candidates(n_trials, -1);
+    std::vector<dtype> next_center_candidates_rand(n_trials, 0);
+    std::vector<dtype> next_center_candidates_potential(n_trials, 0);
+    /* allocate space for the array holding the squared distances to the assigned cluster centers */
+    std::vector<dtype> squared_distances(n_frames, 0);
+
+    /* create the output objects */
+    std::vector<size_t> shape = {k, dim};
+    np_array ret_init_centers(shape);
+    auto init_centers = ret_init_centers.mutable_unchecked();
+    std::memset(init_centers.mutable_data(), 0, init_centers.size() * init_centers.itemsize());
+
+    auto data = np_data.template unchecked<2>();
+
+    /* initialize random device and pick first center randomly */
+    std::default_random_engine generator(random_seed);
+    std::uniform_int_distribution<size_t> uniform_dist(0, n_frames - 1);
+    first_center_index = uniform_dist(generator);
+    /* and mark it as assigned */
+    taken_points[first_center_index] = true;
+    /* write its coordinates into the init_centers array */
+    for (j = 0; j < dim; j++) {
+        init_centers(centers_found, j) = data(first_center_index, j);
+    }
+    /* increase number of found centers */
+    centers_found++;
+    /* perform callback */
+    if (! py::isinstance<py::none>(callback)) {
+        callback();
+    }
+#ifdef USE_OPENMP
+    omp_set_num_threads(n_threads);
+#endif
+
+    /* iterate over all data points j, measuring the squared distance between j and the initial center i: */
+    /* squared_distances[i] = distance(x_j, x_i)*distance(x_j, x_i) */
+    dtype dist_sum = 0.0;
+    #pragma omp parallel for
```

Comment:
I don't think we can parallelize this by frames since the state when processing frame `n` depends on the outcome of processing frames `1...(n-1)`. I think the only true parallelism here to end up with a deterministic outcome is to parallelize the distance calculation to existing centers. 

In reality, anyone who wants to really parallelize this would use the equitemporal initialization, since this is trivially parallelizable.